### PR TITLE
[DataPipe] Unlimited buffer for Forker and Demultiplexer

### LINKED ISSUE #64991: [DataPipe] Forker and Demultiplexer should be able to take unlimited buffer
## 🚀 Feature

Add an option to `Forker` and `Demultiplexer` to use unlimited buffer if `buffer_size` is set to None/-1.

cc @VitalyFedyunin @ejguan

### PR DESCRIPTION
**Pull request by ejguan**
Fixes #64991

Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #64994

When `buffer_size=-1`, use unlimited buffer for Forker and Demultiplexer.

Differential Revision: [D30934362](https://our.internmc.facebook.com/intern/diff/D30934362)

### CODE REVIEW

**Comment by ejguan** [comment]
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30934362).

**Comment by ejguan** [comment]
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30934362).

**Comment by NivekT** [comment]
LGTM!

**Review by NivekT — APPROVED** [review]

**Comment by ejguan** [comment]
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30934362).

**Comment by ejguan** [comment]
@NivekT 
Do we accept `buffer_size=0` for Forker and Demulitplexer?

File: torch/utils/data/datapipes/iter/combining.py [hunk 1]
```diff
@@ -181,7 +182,8 @@ class DemultiplexerIterDataPipe(IterDataPipe):
             classifier_fn: a function that maps values to an integer within the range [0, num_instances - 1] or None
             drop_none: defaults to False, if True, the function will skip over elements classified as None
             buffer_size: this defines the maximum number of inputs that the buffer can hold across all child
-                DataPipes while waiting for their values to be yielded
+                DataPipes while waiting for their values to be yielded.
+                Use -1 for the unlimited buffer
     """
     def __new__(cls, datapipe: IterDataPipe, num_instances: int,
                 classifier_fn: Callable[[T_co], int], drop_none: bool = False, buffer_size: int = 1000):
```
Comment (NivekT) [inline torch/utils/data/datapipes/iter/combining.py:189]: Should we make it unlimited by default? Or a larger default? 

The question would be - how many users would run into memory issues if we make the default larger or unlimited.

**Comment by NivekT** [comment]
> @NivekT
> Do we accept `buffer_size=0` for Forker and Demulitplexer?

@ejguan `buffer_size=0` currently doesn't work for Forker because we always add the incoming element to buffer before popping.

In contrast, Demulitplexer immediately returns if `instance_id == classification`, so `0` works.

In either case, `buffer_size < 0` as a condition for infinite size will work for both DataPipe

File: torch/utils/data/datapipes/iter/combining.py (hunk 1 above)
Comment (ejguan) [reply torch/utils/data/datapipes/iter/combining.py:189]: Let me add a global default buffer size (maybe 100_000) and let all buffers using it. If all buffer unlimited by default, the chance for OOM would be great. So,I would do a larger default for now. 

If more user requests are found, we are able to convert it to unlimited more smoothly (without breaking users' code)

Comment (NivekT) [reply torch/utils/data/datapipes/iter/combining.py:189]: That's a good idea. Perhaps we can have a function that allows the users to set the global default too?

**Comment by ejguan** [comment]
> `buffer_size=0` currently doesn't work for Forker because we always add the incoming element to buffer before popping. We can change this behavior if needed though.
> 
> In contrast, Demulitplexer immediately returns if `instance_id == classification`, so `0` works.

I will take a look to see if we really need to change it. TBH, I think it's fine. For `Forker` with `instance_num=1`, `buffer_size=0` is working.

Then, I need to modify this PR to check if `buffer_size >= 0` rather than `buffer_size = 0`.

File: torch/utils/data/datapipes/iter/combining.py (hunk 1 above)
Comment (ejguan) [reply torch/utils/data/datapipes/iter/combining.py:189]: Sure, then let's do another PR to do the both things:
- Add global default buffer size
- Add a way to set global default

**Comment by VitalyFedyunin** [comment]
We should probably include doc warning, as using unlimited buffer might lead to OOMs at random places.

**Comment by ejguan** [comment]
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30934362).

**Pull request closed without merging**